adapters/algorix/algorix.go

```diff
@@ -0,0 +1,166 @@
+package algorix
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"net/http"
+	"text/template"
+
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	EndpointTemplate template.Template
+}
+
+// Builder builds a new instance of the Foo adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	endpoint, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+	bidder := &adapter{
+		EndpointTemplate: *endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var adapterRequests []*adapters.RequestData
+	var errs []error
+
+	adapterRequest, err := a.makeRequest(request)
+	if err == nil {
+		adapterRequests = append(adapterRequests, adapterRequest)
+	} else {
+		errs = append(errs, err)
+	}
+	return adapterRequests, errs
+}
+
+func (a *adapter) makeRequest(request *openrtb2.BidRequest) (*adapters.RequestData, error) {
+	algorixExt := parseAlgoriXExt(request)
+
+	if algorixExt == nil {
+		return nil, &errortypes.BadInput{Message: "Invalid ExtImpAlgoriX value"}
+	}
+
+	endPoint, err := a.getEndPoint(algorixExt)
+	if err != nil {
+		return nil, err
+	}
+
+	preProcess(request)
+	reqBody, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     endPoint,
+		Body:    reqBody,
+		Headers: headers,
+	}, nil
+}
+
+//parseAlgoriXExt parse AlgoriX Ext
+func parseAlgoriXExt(request *openrtb2.BidRequest) *openrtb_ext.ExtImpAlgorix {
+	var extImpAlgoriX openrtb_ext.ExtImpAlgorix
+	for _, imp := range request.Imp {
+		var extBidder adapters.ExtImpBidder
+		err := json.Unmarshal(imp.Ext, &extBidder)
+		if err != nil {
+			continue
+		}
+		err = json.Unmarshal(extBidder.Bidder, &extImpAlgoriX)
+		if err != nil {
+			continue
+		}
+		return &extImpAlgoriX
+	}
+	return nil
+}
```

Comment:
This algorithm is searching for the first Imp with valid parameters and then using those for the entire request. I have two recommendations:

1 - Return an error if the unmarshal fails for any Imp. This indicates an invalid request and is a good reason to return an error. In practice, this won't happen since the core logic of Prebid Server will perform this check earlier in the request pipeline and return a fatal error. IMHO it's best to match Prebid Server core's behavior here.

2 - Please make it clear by adding a note to your separate documentation PR that only the parameters of the first Imp matter. The rest are ignored by this adapter. Consider adding a json test for this situation as well to provide an example to publishers.

Reply:
We change the code. only check and fetch first Imp extBidder, ignored rest imp.
also we add json test example(two imps and second is invalid, Prebid server adapter will ignored it. and we add this note for Doc PR